src/precompiles/modexp.rs

```diff
@@ -132,24 +89,96 @@ impl Precompile for ModExp<Byzantium> {
             }
         };
 
-        Ok(PrecompileOutput::without_logs(cost, output))
+        Ok(output)
+    }
+}
+
+impl ModExp<Byzantium> {
+    fn mul_complexity(x: u64) -> Result<u64, ExitError> {
+        if x <= 64 {
+            Ok(x * x)
+        } else if x <= 1_024 {
+            Ok(x * x / 4 + 96 * x - 3_072)
+        } else {
+            let (sqroot, overflow) = x.overflowing_mul(x);
+            if overflow {
+                Err(ExitError::OutOfGas)
+            } else {
+                Ok(sqroot / 16 + 480 * x - 199_680)
+            }
+        }
+    }
+}
+
+impl Precompile for ModExp<Byzantium> {
+    fn required_gas(input: &[u8]) -> Result<u64, ExitError> {
+        let base_len = U256::from(&input[0..32]).as_u64();
+        let exp_len = U256::from(&input[32..64]).as_u64();
+        let mod_len = U256::from(&input[64..96]).as_u64();
+
+        let mul = Self::mul_complexity(core::cmp::max(mod_len, base_len))?;
+        let iter_count = Self::calc_iter_count(exp_len, base_len, &input) - 1;
```

Comment:
Sorry, I'm all over the place today :(

Does this underflow in the case the exponent is zero?

Reply:
Good catch! Other implementations just subtracted the 1. So, hmm. Good question. I added all test case though. I think one of them is 0 exponent.

Reply:
The minus 1 is only in the non-zero cases in the spec. Also in the last case, the minus 1 is only on the bit length, not the whole expression.

```
def calculate_iteration_count(exponent_length, exponent):
    iteration_count = 0
    if exponent_length <= 32 and exponent == 0: iteration_count = 0
    elif exponent_length <= 32: iteration_count = exponent.bit_length() - 1
    elif exponent_length > 32: iteration_count = (8 * (exponent_length - 32)) + ((exponent & (2**256 - 1)).bit_length() - 1)
    return max(iteration_count, 1)
```

Reply:
Ah yes, good catch. Fixed now in https://github.com/aurora-is-near/aurora-engine/pull/138/commits/0494f937a2f5cd930846ecdca3278e6a64a14b05.